File: py/Frame.py
from enum import IntEnum
from Utils import Assert


class OpCode(IntEnum):
    OP_NEW_NUM = 0,
    OP_NEW_STR = 1,
    OP_NEW_TRUE = 2,
    OP_NEW_FALSE = 3,
    OP_NEW_NIL = 4,
    OP_NEW_ARRAY = 5,
    OP_NEW_STRUCT = 6,
    OP_GET_VAR = 7,
    OP_SET_VAR = 8,
    OP_DEFINE_VAR = 9,
    OP_GET_INDEX_VAR = 10,
    OP_SET_INDEX_VAR = 11,
    OP_GET_STRUCT_VAR = 12,
    OP_SET_STRUCT_VAR = 13,
    OP_NEG = 14,
    OP_RETURN = 15,
    OP_ADD = 16,
    OP_SUB = 17,
    OP_MUL = 18,
    OP_DIV = 19,
    OP_GREATER = 20,
    OP_LESS = 21,
    OP_GREATER_EQUAL = 22,
    OP_LESS_EQUAL = 23,
    OP_EQUAL = 24,
    OP_NOT_EQUAL = 25,
    OP_NOT = 26,
    OP_OR = 27,
    OP_AND = 28,
    OP_ENTER_SCOPE = 29,
    OP_EXIT_SCOPE = 30,
    OP_JUMP = 30,
    OP_JUMP_IF_FALSE = 31,
    OP_FUNCTION_CALL = 32,


class Frame:
    upFrame: Frame = None
    functionFrames: dict[str, Frame] = {}
    structFrames: dict[str, Frame] = {}
    codes: list[int] = []
    nums: list[float] = []
    strings: list[str] = []

    def __init__(self, upFrame=None) -> None:
        self.upFrame = upFrame

    def AddOpCode(self, code: int) -> None:
        self.codes.append(code)

    def GetOpCodeSize(self) -> int:
        return len(self.codes)

    def AddNum(self, value: float) -> int:
        self.nums.append(value)
        return len(self.nums)-1

    def AddString(self, value: str) -> int:
        self.strings.append(value)
        return len(self.strings)-1

    def AddFunctionFrame(self, name: str, frame: Frame) -> None:
        if self.functionFrames.get(name) != None:
            Assert("Redefinition function:"+name)
        self.functionFrames[name] = frame

    def GetFunctionFrame(self, name: str) -> Frame:
        if self.functionFrames.get(name) != None:
            return self.functionFrames.get(name)
        elif self.upFrame != None:
            return self.upFrame.GetFunctionFrame(name)

    def HasFunctionFrame(self, name: str) -> bool:
        for key in self.functionFrames:
            if key == name:
                return True
        if self.upFrame != None:
            return self.upFrame.HasFunctionFrame(name)
        return False

    def AddStructFrame(self, name: str, frame: Frame) -> None:
        if self.structFrames.get(name) != None:
            Assert("Redefinition function:"+name)
        self.functstructFramesionFrames[name] = frame

    def GetStructFrame(self, name: str) -> Frame:
        if self.structFrames.get(name) != None:
            return self.structFrames.get(name)
        elif self.upFrame != None:
            return self.upFrame.GetFunctionFrame(name)

    def HasStructFrame(self, name: str) -> bool:
        for key in self.structFrames:
            if key == name:
                return True
        if self.upFrame != None:
            return self.upFrame.HasStructFrame(name)
        return False

    def Clear(self) -> None:
        self.upFrame: Frame = None
        self.functionFrames: dict[str, Frame] = {}
        self.structFrames: dict[str, Frame] = {}
        self.codes: list[int] = []
        self.nums: list[float] = []
        self.strings: list[str] = []

    def Stringify(self, depth: int = 0) -> str:
        interval: str = ""
        for i in range(depth):
            interval += "\t"

        result: str = ""

        for key, value in self.structFrames.items():
            result += ("%sFrame %d:\n" % (interval, key))
            result += value.Stringify(depth+1)

        for key, value in self.functionFrames.items():
            result += ("%sFrame %d:\n" % (interval, key))
            result += value.Stringify(depth+1)

        result += ("%sOpCodes:\n" % interval)

        for i in range(len(self.codes)):
            if self.codes[i] == OpCode.OP_RETURN:
                result += "%s\t%08d     OP_RETURN\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_NEW_NUM:
                result += "%s\t%08d     OP_NEW_NUM     %d\n" % (
                    interval, i, self.nums[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_NEW_STR:
                result += "%s\t%08d     OP_NEW_STR     %s\n" % (
                    interval, i, self.strings[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_NEW_TRUE:
                result += "%s\t%08d     OP_NEW_TRUE\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_NEW_FALSE:
                result += "%s\t%08d     OP_NEW_FALSE\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_NEW_NIL:
                result += "%s\t%08d     OP_NEW_NIL\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_NEW_STRUCT:
                result += "%s\t%08d     OP_NEW_STRUCT     %s\n" % (
                    interval, i, self.strings[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_NEG:
                result += "%s\t%08d     OP_NEG\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_ADD:
                result += "%s\t%08d     OP_ADD\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_SUB:
                result += "%s\t%08d     OP_SUB\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_MUL:
                result += "%s\t%08d     OP_MUL\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_DIV:
                result += "%s\t%08d     OP_DIV\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_GREATER:
                result += "%s\t%08d     OP_GREATER\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_LESS:
                result += "%s\t%08d     OP_LESS\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_GREATER_EQUAL:
                result += "%s\t%08d     OP_GREATER_EQUAL\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_LESS_EQUAL:
                result += "%s\t%08d     OP_LESS_EQUAL\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_EQUAL:
                result += "%s\t%08d     OP_EQUAL\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_NOT:
                result += "%s\t%08d     OP_NOT\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_NOT_EQUAL:
                result += "%s\t%08d     OP_NOT_EQUAL\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_AND:
                result += "%s\t%08d     OP_AND\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_OR:
                result += "%s\t%08d     OP_OR\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_GET_VAR:
                result += "%s\t%08d     OP_GET_VAR     %s\n" % (
                    interval, i, self.strings[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_DEFINE_VAR:
                result += "%s\t%08d     OP_DEFINE_VAR     %s\n" % (
                    interval, i, self.strings[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_SET_VAR:
                result += "%s\t%08d     OP_SET_VAR     %s\n" % (
                    interval, i, self.strings[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_NEW_ARRAY:
                result += "%s\t%08d     OP_NEW_ARRAY     %d\n" % (
                    interval, i, self.nums[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_GET_INDEX_VAR:
                result += "%s\t%08d     OP_GET_INDEX_VAR\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_SET_INDEX_VAR:
                result += "%s\t%08d     OP_SET_INDEX_VAR\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_GET_STRUCT_VAR:
                result += "%s\t%08d     OP_GET_STRUCT_VAR     %s\n" % (
                    interval, i, self.strings[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_SET_STRUCT_VAR:
                result += "%s\t%08d     OP_SET_STRUCT_VAR     %s\n" % (
                    interval, i, self.strings[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_ENTER_SCOPE:
                result += "%s\t%08d     OP_ENTER_SCOPE\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_EXIT_SCOPE:
                result += "%s\t%08d     OP_EXIT_SCOPE\n" % (interval, i)
            elif self.codes[i] == OpCode.OP_JUMP:
                result += "%s\t%08d     OP_JUMP     %d\n" % (
                    interval, i, self.nums[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_JUMP_IF_FALSE:
                result += "%s\t%08d     OP_JUMP_IF_FALSE     %d\n" % (
                    interval, i, self.nums[self.codes[i+1]])
                i = i+1
            elif self.codes[i] == OpCode.OP_FUNCTION_CALL:
                result += "%s\t%08d     OP_FUNCTION_CALL     %d\n" % (
                    interval, i, self.strings[self.codes[i+1]])
                i = i+1
            else:
                result += "%s\t%08d     UNKNOWN\n" % (interval, i)
        return result

File: Frame.cpp
#include "Frame.h"
#include "Utils.h"

Frame::Frame()
{
}

Frame::~Frame()
{
	Clear();
}
void Frame::AddOpCode(uint64_t code)
{
	m_Codes.emplace_back(code);
}

uint64_t Frame::GetOpCodeSize() const
{
	return m_Codes.size();
}

uint64_t Frame::AddNum(double value)
{
	m_Nums.emplace_back(value);
	return m_Nums.size() - 1;
}

uint64_t Frame::AddString(std::string_view value)
{
	m_Strings.emplace_back(value);
	return m_Strings.size() - 1;
}

std::vector<double> &Frame::GetNums()
{
	return m_Nums;
}

uint64_t Frame::AddFunctionFrame(Frame frame)
{
	m_FunctionFrames.emplace_back(frame);
	return m_FunctionFrames.size() - 1;
}

Frame Frame::GetFunctionFrame(uint64_t idx)
{
		return m_FunctionFrames[idx];
}

bool Frame::HasFunctionFrame(uint64_t idx)
{
	if (idx >= 0 || idx < m_FunctionFrames.size())
		return true;
	else
		return false;
}

std::string Frame::Stringify(int depth)
{
	std::string interval;
	for (size_t i = 0; i < depth; ++i)
		interval += "\t";

#define SINGLE_INSTR_STRINGIFY(op) \
	result << interval << "\t" << std::setfill('0') << std::setw(8) << i << "     " << (#op) << "\n"

#define CONSTANT_INSTR_STRINGIFY(op, vec) \
	result << interval << "\t" << std::setfill('0') << std::setw(8) << i << "     " << (#op) << "     " << vec[m_Codes[++i]] << "\n"

	std::stringstream result;

	for (size_t i = 0; i < m_FunctionFrames.size(); ++i)
	{
		result << interval << "Frame " << i << ":\n";
		result << m_FunctionFrames[i].Stringify(depth + 1);
	}

	result << interval << "OpCodes:\n";

	for (size_t i = 0; i < m_Codes.size(); ++i)
	{
		switch (m_Codes[i])
		{
		case OP_RETURN:
			SINGLE_INSTR_STRINGIFY(OP_RETURN);
			break;
		case OP_NEW_NUM:
			CONSTANT_INSTR_STRINGIFY(OP_NEW_NUM, m_Nums);
			break;
		case OP_NEW_STR:
			CONSTANT_INSTR_STRINGIFY(OP_NEW_STR, m_Strings);
			break;
		case OP_NEW_TRUE:
			SINGLE_INSTR_STRINGIFY(OP_NEW_TRUE);
			break;
		case OP_NEW_FALSE:
			SINGLE_INSTR_STRINGIFY(OP_NEW_FALSE);
			break;
		case OP_NEW_NIL:
			SINGLE_INSTR_STRINGIFY(OP_NEW_NIL);
			break;
		case OP_NEG:
			SINGLE_INSTR_STRINGIFY(OP_NEG);
			break;
		case OP_ADD:
			SINGLE_INSTR_STRINGIFY(OP_ADD);
			break;
		case OP_SUB:
			SINGLE_INSTR_STRINGIFY(OP_SUB);
			break;
		case OP_MUL:
			SINGLE_INSTR_STRINGIFY(OP_MUL);
			break;
		case OP_DIV:
			SINGLE_INSTR_STRINGIFY(OP_DIV);
			break;
		case OP_GREATER:
			SINGLE_INSTR_STRINGIFY(OP_GREATER);
			break;
		case OP_LESS:
			SINGLE_INSTR_STRINGIFY(OP_LESS);
			break;
		case OP_GREATER_EQUAL:
			SINGLE_INSTR_STRINGIFY(OP_GREATER_EQUAL);
			break;
		case OP_LESS_EQUAL:
			SINGLE_INSTR_STRINGIFY(OP_LESS_EQUAL);
			break;
		case OP_EQUAL:
			SINGLE_INSTR_STRINGIFY(OP_EQUAL);
			break;
		case OP_NOT:
			SINGLE_INSTR_STRINGIFY(OP_NOT);
			break;
		case OP_NOT_EQUAL:
			SINGLE_INSTR_STRINGIFY(OP_NOT_EQUAL);
			break;
		case OP_AND:
			SINGLE_INSTR_STRINGIFY(OP_AND);
			break;
		case OP_OR:
			SINGLE_INSTR_STRINGIFY(OP_OR);
			break;
		case OP_GET_VAR:
			CONSTANT_INSTR_STRINGIFY(OP_GET_VAR, m_Strings);
			break;
		case OP_DEFINE_VAR:
			CONSTANT_INSTR_STRINGIFY(OP_DEFINE_VAR, m_Strings);
			break;
		case OP_SET_VAR:
			CONSTANT_INSTR_STRINGIFY(OP_SET_VAR, m_Strings);
			break;
		case OP_NEW_ARRAY:
			CONSTANT_INSTR_STRINGIFY(OP_NEW_ARRAY, m_Nums);
			break;
		case OP_NEW_FUNCTION:
			CONSTANT_INSTR_STRINGIFY(OP_NEW_FUNCTION, m_Nums);
			break;
		case OP_GET_INDEX_VAR:
			SINGLE_INSTR_STRINGIFY(OP_GET_INDEX_VAR);
			break;
		case OP_SET_INDEX_VAR:
			SINGLE_INSTR_STRINGIFY(OP_SET_INDEX_VAR);
			break;

		case OP_ENTER_SCOPE:
			SINGLE_INSTR_STRINGIFY(OP_ENTER_SCOPE);
			break;
		case OP_EXIT_SCOPE:
			SINGLE_INSTR_STRINGIFY(OP_EXIT_SCOPE);
			break;
		case OP_JUMP:
			CONSTANT_INSTR_STRINGIFY(OP_JUMP, m_Nums);
			break;
		case OP_JUMP_IF_FALSE:
			CONSTANT_INSTR_STRINGIFY(OP_JUMP_IF_FALSE, m_Nums);
			break;
		case OP_FUNCTION_CALL:
			CONSTANT_INSTR_STRINGIFY(OP_FUNCTION_CALL,m_Strings);
			break;
		case OP_REF:
			SINGLE_INSTR_STRINGIFY(OP_REF);
			break;
		default:
			SINGLE_INSTR_STRINGIFY(UNKNOWN);
			break;
		}
	}

	return result.str();
}
void Frame::Clear()
{
	std::vector<uint64_t>().swap(m_Codes);
	std::vector<double>().swap(m_Nums);
	std::vector<std::string>().swap(m_Strings);

	for (auto funcFrame : m_FunctionFrames)
		funcFrame.Clear();

	std::vector<Frame>().swap(m_FunctionFrames);
}